### PR TITLE
Fix race condition and data race in FrameInfoContainsValidWidthAndHeight.

### PR DESCRIPTION
**Pull request by chinmaygarde**
The data race: gl_surface_fbo_frame_infos_ in the test context to was appended
to on the raster task runner but read on the platform task runner. This is now
sidestepped by using a callback and pushing that responsibility to the test.
Setting the callback is guarded behind a mutex.

The race condition: The assertions were previously run when the UI thread was
done generating the frames. However, the assertions were run on the results
collected on the raster thread in response the frame requests from UI thread.
Just run the assertions on the raster thread directly.

Fixes https://github.com/flutter/flutter/issues/64344

### CODE REVIEW

File: shell/platform/embedder/tests/embedder_test_context.h [hunk 1]
```diff
@@ -79,8 +80,20 @@ class EmbedderTestContext {
...
-  std::vector<FlutterFrameInfo> GetGLFBOFrameInfos();
+  using GLGetFBOCallback = std::function<void(FlutterFrameInfo frame_info)>;
+
+  //----------------------------------------------------------------------------
+  /// @brief      Sets a callback that will be invoked (on the raster task
+  ///             runner) when the engine asks the embedder for a new FBO ID at
+  ///             the updated size.
+  ///
+  /// @attention  The callback will be invoked on the raster task runner. The
+  ///             callback can be set tests host thread.
```
Comment (zanderso) [inline shell/platform/embedder/tests/embedder_test_context.h:91]: Is there a missing word here?

File: shell/platform/embedder/tests/embedder_test_context.h [hunk 2]
```diff
@@ -79,8 +80,20 @@ class EmbedderTestContext {
...
+  ///
+  /// @param[in]  callback  The callback to set. The previous callback will be
+  ///                       un-registered.
+  ///
+  void SetGLGetFBOCallback(GLGetFBOCallback callback);
```
Comment (zanderso) [inline shell/platform/embedder/tests/embedder_test_context.h:96]: Does this have any use outside of testing? Is there any way to make a method visible only for a test?

Comment (chinmaygarde) [reply shell/platform/embedder/tests/embedder_test_context.h:96]: `embedder/tests/embedder_test_context.h` is a test only class.

File: shell/platform/embedder/tests/embedder_test_context.h (hunk 1 above)
Comment (chinmaygarde) [reply shell/platform/embedder/tests/embedder_test_context.h]: Fixed typo.

**Pull request merged**